// adafruit_pixel/adafruit_pixel.js
var spi = require('spi');
var Easing = require('easing');

var STEPS = 25;
var DURATION_SCALE = 1000/STEPS;

function Throb(setter, pixels, start_color, end_color, duration, options)
{
    this.setter = setter;
    this.pixels = pixels;
    this.start_color = start_color;
    this.end_color = end_color;
    this.duration = duration;
    if (options !== undefined && "easing" in options) {
        this.easing = Easing(STEPS, options.easing, {
            endToEnd:true
        });
    } else {
        this.easing = Easing(STEPS, 'linear', {
            endToEnd:true
        });
    }
    this.step = 0;
    this.running = false;
}

Throb.prototype.start = function() {
    console.log(this.easing);
    this.running = true;
    this.tick();
}

Throb.prototype.stop = function() {
    this.running = false;
}

Throb.prototype.calculate_single = function(start_value, end_value) {
    var retval = start_value + (end_value-start_value)*this.easing[this.step];
    return retval;
}

Throb.prototype.calculate_rgb = function() {
    var r = this.calculate_single(this.start_color[0],
                                  this.end_color[0]);
    var g = this.calculate_single(this.start_color[1],
                                  this.end_color[1]);
    var b = this.calculate_single(this.start_color[2],
                                  this.end_color[2]);

    return [r,g,b];
}

Throb.prototype.tick = function() {
    var self = this;
    rgb = this.calculate_rgb();
    // Loop through all of the pixels in pixels, and call set
    this.pixels.forEach(function(element, index, array) {
        self.setter.set(element, rgb[0], rgb[1], rgb[2]);
    });
    self.setter.sync();

    this.step = (this.step+=1)%STEPS;
    if (this.running) {
        setTimeout(function() { self.tick(); },
            this.duration*DURATION_SCALE);
    }
}

function Pixel(deviceName, num_pixels) {
    console.log("num_pixels: " + num_pixels);
    this.enable = true;
    this.num_pixels = num_pixels;
    this.StartPixel = 4;
    this.EndPixel = this.StartPixel+this.num_pixels*4;
    this.DataEnd = this.EndPixel+8;

    this.pixel_buffer = new Buffer(this.DataEnd);
    this.initPixels(this.pixel_buffer ,255 ,0 ,0);
    // console.log(this.pixel_buffer);

    this.off_buffer = new Buffer(this.DataEnd);
    this.initPixels(this.off_buffer ,255 ,0 ,0);
    // console.log(this.off_buffer);

    this.rd_buffer = new Buffer(this.DataEnd);
    this.initPixels(this.rd_buffer ,255 ,0 ,0);
    // console.log(this.rd_buffer);
    
    console.log("device0: " + deviceName);
    if(this.enable)this.device = new spi.Spi(deviceName);
    //this.device.open(deviceName, {
    //    "mode": spi.MODE[0],
    //    "chipSelect": spi.CS['none'],
    //    "maxSpeed": 1000000
    //});
    
    // this.pixel_buffer.fill(0);
    // this.off_buffer.fill(0);
    if(this.enable)this.device.write(this.pixel_buffer, this.rd_buffer);
    // console.log(this.rd_buffer);
    this.animate = null;
};
Pixel.prototype.initPixels = function(buffer , r , g , b) {
    
    buffer[0] = 0;
    buffer[1] = 0;
    buffer[2] = 0;
    buffer[3] = 0;
    for(var i = 1; i < this.num_pixels+1; i++) {
        buffer[i*4] = this.makeflag(r,g,b);
        buffer[i*4+1] = r;
        buffer[i*4+2] = g;
        buffer[i*4+3] = b;
    }

    buffer[this.EndPixel] = 0;
    buffer[this.EndPixel+1] = 0;
    buffer[this.EndPixel+2] = 0;
    buffer[this.EndPixel+3] = 0;
    buffer[this.EndPixel+4] = 0;
    buffer[this.EndPixel+5] = 0;
    buffer[this.EndPixel+6] = 0;
    buffer[this.EndPixel+7] = 0;
    // console.log("------------------------------------------------------------------------");
    // for(var i = 0 ; i < this.num_pixels+3 ;i++)
    // {
    //     console.log(i+": " + buffer[i*4] +" "+ buffer[i*4+1] +" "+ buffer[i*4+2] +" "+ buffer[i*4+3]);
    // }
    // console.log("------------------------------------------------------------------------");
   
};
Pixel.prototype.off = function() {
    if(this.enable)this.device.write(this.off_buffer);
};

Pixel.prototype.sync = function() {
    if(this.enable)this.device.write(this.pixel_buffer, this.rd_buffer);
};

Pixel.prototype.all = function(r,g,b) {
    for(var i = 1; i < this.num_pixels+1; i++) {
        this.set(i, r, g, b);
    }
}

Pixel.prototype.clear = function() {
    this.pixel_buffer.fill(0);
}
Pixel.prototype.fill = function(buffer) {
    for(var i = 1; i < this.num_pixels+1; i++) {
        this.set(i, 0,0,0);
    }
}
Pixel.prototype.set = function(pixel, r, g, b) {
    this.pixel_buffer[pixel*4] = this.makeflag(r,g,b);
    this.pixel_buffer[pixel*4+1] = b;
    this.pixel_buffer[pixel*4+2] = g;
    this.pixel_buffer[pixel*4+3] = r;
}

Pixel.prototype.throb = function(pixels, start_color, end_color, duration, options) {
    this.stop();
    this.animate = new Throb(this, pixels, start_color, end_color, duration, options);
    this.animate.start();
}

Pixel.prototype.stop = function() {
    if (this.animate !== null) {
        this.animate.stop();
        this.animate = null;
    }
}
Pixel.prototype.makeflag = function(r, g, b) {
    flag = 0;
    
    flag =  (r&0xc0)>>6;
    flag |= (g&0xc0)>>4;
    flag |= (b&0xc0)>>2;
    
    return ~flag;
}

module.exports.Pixel = Pixel;
